pubsub/iterator_test.go

```diff
@@ -618,3 +619,207 @@ func TestPingStreamAckDeadline(t *testing.T) {
 	}
 	iter.eoMu.RUnlock()
 }
+
+func compareCompletedRetryLengths(t *testing.T, completed, retry []*AckResult, wantCompleted, wantRetry int) {
+	if l := len(completed); l != wantCompleted {
+		t.Errorf("completed slice length is %d, want %d\n", l, wantCompleted)
+	}
+	if l := len(retry); l != wantRetry {
+		t.Errorf("retry slice length is %d, want %d\n", l, wantRetry)
+	}
+}
+
+func TestExactlyOnceProcessRequests(t *testing.T) {
+	ctx := context.Background()
+
+	t.Run("NoResults", func(t *testing.T) {
+		// If the ackResMap is nil, then the resulting slices should be empty.
+		// nil maps here behave the same as if they were empty maps.
+		completed, retry := processResults(nil, nil, nil)
+		compareCompletedRetryLengths(t, completed, retry, 0, 0)
+	})
+
+	t.Run("NoErrorsNilAckResult", func(t *testing.T) {
+		// No errors so request should be completed even without an AckResult.
+		ackReqMap := map[string]*AckResult{
+			"ackID": nil,
+		}
+		completed, retry := processResults(nil, ackReqMap, nil)
+		compareCompletedRetryLengths(t, completed, retry, 1, 0)
+	})
+
+	t.Run("NoErrors", func(t *testing.T) {
+		// No errors so AckResult should be completed with success.
+		r := ipubsub.NewAckResult()
+		ackReqMap := map[string]*AckResult{
+			"ackID1": r,
+		}
+		completed, retry := processResults(nil, ackReqMap, nil)
+		compareCompletedRetryLengths(t, completed, retry, 1, 0)
+
+		// We can obtain the AckStatus from AckResult if results are completed.
+		s, err := r.Get(ctx)
+		if err != nil {
+			t.Errorf("got err from AckResult: %v", err)
+		}
+		if s != AcknowledgeStatusSuccess {
+			t.Errorf("expected AcknowledgeStatusSuccess, got %v", s)
```

Comment:
nit: phrase as "got %v, want AcknowledgeStatusSuccess". I think this happens more below as well.